aptos-move/block-executor/src/proptest_types/baseline.rs

```diff
@@ -0,0 +1,287 @@
+    SkipRest,
+    GasLimitExceeded,
+}
+
+/// Sequential baseline of execution result for dummy transaction, containing a vector
+/// of BaselineValues for the reads of the (latest incarnation of the dummy) transaction.
+/// The size of the vector should be equal to the size of the block if the block execution
+/// was successful. Otherwise, it is the index of a transaction where the block execution
+/// stopped, e.g. due to gas limit, abort, or reconfiguration (skip rest status). It also
+/// contains resolved values for each of the deltas produced by the dummy transaction.
+///
+/// For both read_values and resolved_deltas the keys are not included because they are
+/// in the same order as the reads and deltas in the Transaction::Write.
+pub(crate) struct BaselineOutput<V> {
+    status: BaselineStatus,
+    read_values: Vec<Result<Vec<BaselineValue<V>>, ()>>,
+    resolved_deltas: Vec<Result<Vec<u128>, ()>>,
+}
+
+impl<V: Debug + Clone + PartialEq + Eq + TransactionWrite> BaselineOutput<V> {
+    /// Must be invoked after parallel execution to have incarnation information set and
+    /// work with dynamic read/writes.
+    pub(crate) fn generate<K: Hash + Clone + Eq>(
+        txns: &[MockTransaction<K, V>],
+        maybe_block_gas_limit: Option<u64>,
+    ) -> Self {
+        let mut current_world = HashMap::<K, BaselineValue<_>>::new();
+        let mut accumulated_gas = 0;
+
+        let mut status = BaselineStatus::Success;
+        let mut read_values = vec![];
+        let mut resolved_deltas = vec![];
+        for txn in txns.iter() {
+            match txn {
+                MockTransaction::Abort => {
+                    status = BaselineStatus::Aborted;
+                    break;
+                },
+                MockTransaction::SkipRest => {
+                    // In executor, SkipRest skips from the next index. Test assumes it's an empty
+                    // transaction, so create a successful empty reads and deltas.
+                    read_values.push(Ok(vec![]));
+                    resolved_deltas.push(Ok(vec![]));
+
+                    status = BaselineStatus::SkipRest;
+                    break;
+                },
+                MockTransaction::Write {
+                    incarnation,
+                    writes,
+                    deltas,
+                    reads,
+                    gas,
+                } => {
+                    // Determine the read-, write-, and delta- sets of the latest incarnation of
+                    // the transaction. The index is based on the value of the incarnation counter
+                    // prior to the fetch_add during the last execution.
+                    let incarnation = incarnation.load(Ordering::SeqCst);
+                    // Ensuring invariants for the correctness of testing itself.
+                    assert!(!reads.is_empty());
+                    assert!(!writes.is_empty());
+                    assert_eq!(writes.len(), deltas.len());
```

Comment:
Why same length of writes and deltas? Is this because we generate them together?

Reply:
this was stupid, we had a number of different dynamic incarnation behaviors that different incarnations could produce (to test the fact that when we re-execute, it can behave differently). But it was generating up to that number of different behaviors, and as an artifact of generating writes and deltas together, these were supposed to be equal.

I just did a refactoring here, fixed the number of alternatives for everything, and wrapped behavior in a type so it can be Vec<behavior for incarnation> enforcing that the lens are going to be the same this way.

Reply:
thanks for bringing it up, led to cleaning up a lot more mess.